lib/SILOptimizer/Mandatory/TFUtilities.cpp

```diff
@@ -1899,6 +1921,57 @@ SILLocation tf::getUserSourceLocation(SILInstruction *inst) {
   return getUserSourceLocation(inst->getDebugLocation());
 }
 
+/// Create a "Const" tensor operation containing the specified scalars, with the
+/// specified shape and elementType (setting dtype).  The resultType is the
+/// TensorHandle type to produce.
+GraphOperationInst *
+tf::createConstTensor(Type elementType, SymbolicValue scalars,
+                      SymbolicValue shape, SILType resultType, SILLocation loc,
+                      GraphGlobalConfiguration &deviceConfig, SILBuilder &B) {
+  auto &context = B.getASTContext();
+  auto &allocator = context.getAllocator();
+
+  // Ensure that the array of initializer values is in the module's allocator.
+  scalars = scalars.cloneInto(allocator);
+  shape = shape.cloneInto(allocator);
+  assert(scalars.getKind() == SymbolicValue::Array &&
+         shape.getKind() == SymbolicValue::Array &&
+         "expected array constants for scalars and shape");
+
+  SmallVector<GraphOperationAttribute, 8> attributes;
+
+  // Add a dtype attribute for the array element.
+  attributes.push_back({
+    context.getIdentifier("dtype"),
+    SymbolicValue::getMetatype(elementType->getCanonicalType())
+  });
+
+  // Add an attribute for the value$tensor attribute.
+  auto tensorSuffix =
+  SILTensorOpInfo::getOperandClassSuffix(SILTensorOpInfo::OperandClass::Tensor);
+  attributes.push_back({
+    context.getIdentifier(std::string("value") + tensorSuffix), scalars
+  });
+
+  // Add the value$shape attribute.
+  auto shapeSuffix =
+  SILTensorOpInfo::getOperandClassSuffix(SILTensorOpInfo::OperandClass::Shape);
+  attributes.push_back({
+    context.getIdentifier(std::string("value") + shapeSuffix), shape
+  });
+
+  // All graph_op's get a device.
+  attributes.push_back({
+    context.getIdentifier(DEVICE_ATTR),
+    SymbolicValue::getString(getDeviceString(deviceConfig.primaryDeviceType),
```

Comment:
use ALL_DEVICES here?

Reply:
This is intended to be usable by multiple clients.  Hard coding ALL_DEVICES doesn't seem general... wouldn't that prevent it from being useful to lowerGraph and other things that need it?

Reply:
We chatted, I know what to do :-)